Hello!

My name is Karina. Today I'm reviewing your project. Sorry for such a long waiting.

Firstly, I'll only mark mistakes, if you have them. I'll give you oppotunity to fix it yourself. Of course, I'll add details if where are any difficulties.

Good luck!

Please respond to my comments and mark changes using color codes. 
Below you will find my comments - please do not change them.
You can find my comments in green, yellow or red boxes like this:

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Success. Everything is done succesfully.
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Remarks. Some recommendations.
</div>

<div class="alert alert-block alert-danger">

<b>Reviewer's comment</b> <a class="tocSkip"></a>

Needs fixing. The block requires some corrections. Work can't be accepted with the red comments.
</div>

You can answer me by using this:

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>

Text here)

</div>

# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building the **credit score** of a potential customer. The **credit score** is used to evaluate the ability of a potential borrower to repay their loan.


The purpose of this project is to explore the dataset provided with data of the bank customers and find out if a customer’s marital status, number of children and some othe indicatives have an impact on whether they will default on a loan. The project relults will be reported for a bank’s loan division. 

1. Is there a connection between having kids and repaying a loan on time?
2. Is there a connection between marital status and repaying a loan on time?
3. Is there a connection between income level and repaying a loan on time?
4. How do different loan purposes affect on-time loan repayment?

## Open the data file and have a look at the general information. 


In [1]:
# Loading all the libraries
import pandas as pd
import numpy as np

# Load the data
data_base = pd.read_csv('/datasets/credit_scoring_eng.csv')

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

   
Correct data loading!
    
</div>    

## Task 1. Data exploration

**Description of the data**
- `children` - the number of children in the family
- `days_employed` - work experience in days
- `dob_years` - client's age in years
- `education` - client's education
- `education_id` - education identifier
- `family_status` - marital status
- `family_status_id` - marital status identifier
- `gender` - gender of the client
- `income_type` - type of employment
- `debt` - was there any debt on loan repayment
- `total_income` - monthly income
- `purpose` - the purpose of obtaining a loan

First we will explore our data in general: to find out the dataframe size and try to check for potential issues with the data throug the first rows. 

In [2]:
# Let's see how many rows and columns our dataset has
data_base.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Our database consists of  21525 entries and 12 columns. Columns names looks fine in terms of sintaxis and data type, however, taking into consideration desription of the data for better understanding it would be better to rename 'dob_years' into 'age', 'total_income' into 'monthly_income' and 'debt' into 'default'.
Two columns have missing data: 'days_employed' and 'total_income' and their datatype shoul be changed to integers, as work experience counts in whole days as well as monthly income usually refers as integers.  

In [3]:
# let's print the first 15 rows
display(data_base.tail(15))

children  days_employed  dob_years            education  education_id  \
21510         2            NaN         28  secondary education             1   
21511         0    -612.569129         29    bachelor's degree             0   
21512         0    -165.377752         26    bachelor's degree             0   
21513         0   -1166.216789         35  secondary education             1   
21514         0    -280.469996         27         some college             2   
21515         1    -467.685130         28  secondary education             1   
21516         0    -914.391429         42    bachelor's degree             0   
21517         0    -404.679034         42    bachelor's degree             0   
21518         0  373995.710838         59  SECONDARY EDUCATION             1   
21519         1   -2351.431934         37      graduate degree             4   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
21510            married                 0      F    employee     0   
21511  civil partnership                 1      F    employee     1   
21512          unmarried                 4      M    business     0   
21513            married                 0      F    employee     0   
21514          unmarried                 4      M    business     0   
21515            married                 0      F    employee     1   
21516            married                 0      F    business     0   
21517  civil partnership                 1      F    business     0   
21518            married                 0      F     retiree     0   
21519           divorced                 3      M    employee     0   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                           purpose  
21510           NaN                      car purchase  
21511     22410.956   buying property for renting out  
21512     23568.233  to get a supplementary education  
21513     40157.783             purchase of the house  
21514     56958.145               building a property  
21515     17517.812                to become educated  
21516     51649.244          purchase of my own house  
21517     28489.529                 buying my own car  
21518     24618.344                 purchase of a car  
21519     18551.846        buy commercial real estate  
21520     35966.698              housing transactions  
21521     24959.969                 purchase of a car  
21522     14347.610                          property  
21523     39054.888                 buying my own car  
21524     13127.587                      to buy a car

In [4]:
# Let's have a general look at the dataset 
print(data_base.describe())

           children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000

There are negative and some extraodinary high (20) values in 'children' column.

The 'dob_years' column contains 0 age row(s) - this issue should be explored.

The 'days_employed' column partially consist of negative values which are impossible in our case. Also it contains some impossible values like 340266 days for row with index=4 or maximum  value of 401750 days (equals to 1100 years), both are many times over the human being abilities. 

The 'dob_years' columns contains a row(s) with 0 age.

The 'education' column containes values both in upper and lower cases. This should be corrected in order to avoid duplications.

At a glance at the dataset, it looks like the rows where 'days_employed' is missing tend to be missing a value in the 'total_income' column. We want to check later if this is the case for all rows where 'days_employed' is missing.



In [5]:
# Get info on data
print(data_base.isna().sum())

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64


Equal amounts of missing values were found in 'days_employed' and 'total_income' columns (2174). We must check if each row containing missing value in one column also contains missing value in another as we can see in the row with index=12.

All other columns look fine.

In [6]:
# Renaming columns for best match Description of the data
data_base = data_base.rename(columns = {'dob_years': 'age', 'total_income' : 'monthly_income', 'debt' : 'default'})

Now we will check if all columns renamed according to Description of data.

In [7]:
print(data_base.info())
raw_data = len(data_base)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   age               21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   default           21525 non-null  int64  
 10  monthly_income    19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None


Now when everyting is fixed with columns names we can try to prove the hypothesis that missing values are located symmetrically in columns 'days_employed' and 'monthly_income'. 

In [8]:
# Creating new DataDrame. wich contents only rows with NaN values in 'monthly_income' column
is_null = data_base[data_base['monthly_income'].isna()]

#Cheking if new DataFrame also contains NaN values in 'days_employed' column
is_null.info()
print()

# Cheking if the days_emplyed values have NaN in the rows, where monthly_income are NaN
print(is_null['days_employed'].unique())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2174 entries, 12 to 21510
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          2174 non-null   int64  
 1   days_employed     0 non-null      float64
 2   age               2174 non-null   int64  
 3   education         2174 non-null   object 
 4   education_id      2174 non-null   int64  
 5   family_status     2174 non-null   object 
 6   family_status_id  2174 non-null   int64  
 7   gender            2174 non-null   object 
 8   income_type       2174 non-null   object 
 9   default           2174 non-null   int64  
 10  monthly_income    0 non-null      float64
 11  purpose           2174 non-null   object 
dtypes: float64(2), int64(5), object(5)
memory usage: 220.8+ KB

[nan]


In [9]:
# Cheking the percentage of missing values in the whole dataset.
print((len(is_null) / len(data_base))*100)

10.099883855981417


Missing values in columns 'days_emploid' and 'monthly_income' looks perfectly symmetric. The size of the DataFrame filtered on NaN condition (2174) correspondends to the amount of missing values, established before. It may be the result of some technical issue in the whole bunch of entries. Later on these NaN values will be replaced with the mean or median values for the appropriate category.

**Intermediate conclusion**

Missing values in columns 'days_emploid' and 'monthly_income' looks perfectly symmetric. The size of the DataFrame filtered on NaN condition (2174) correspondends to the amount of missing values, established before. This case may be the result of some technical issue in the whole bunch of entries. Later on these NaN values will be replaced with the mean or median values for the appropriate category.

Total amount of missing values is about 10% of the whole dataset which is considerably large amount. We need to fill it 

In [10]:
# Checking distribution of missing values between categorical values

print('Distribution of missing values between income_type')
print(data_base[data_base['monthly_income'].isna()]['income_type'].value_counts().sort_index())
print()
print('Distribution of values in the whole dataset')
print(data_base.groupby('income_type')['monthly_income'].count().sort_index())
print()
print('Distribution of missing values between gender')
print(data_base[data_base['monthly_income'].isna()]['gender'].value_counts().sort_index())
print()
print('Distribution of values in the whole dataset')
print(data_base.groupby('gender')['monthly_income'].count().sort_index())
print()
print('Distribution of missing values between education')
print(data_base[data_base['monthly_income'].isna()]['education'].value_counts().sort_index())
print()
print('Distribution of values in the whole dataset')
print(data_base.groupby('education')['monthly_income'].count().sort_index().sort_index())
print()
print('Distribution of missing values between purposes')
print(data_base[data_base['monthly_income'].isna()]['purpose'].value_counts().sort_index())
print()
print('Distribution of values in the whole dataset')
print(data_base.groupby('purpose')['monthly_income'].count().sort_index())


Distribution of missing values between income_type
business          508
civil servant     147
employee         1105
entrepreneur        1
retiree           413
Name: income_type, dtype: int64

Distribution of values in the whole dataset
income_type
business                        4577
civil servant                   1312
employee                       10014
entrepreneur                       1
paternity / maternity leave        1
retiree                         3443
student                            1
unemployed                         2
Name: monthly_income, dtype: int64

Distribution of missing values between gender
F    1484
M     690
Name: gender, dtype: int64

Distribution of values in the whole dataset
gender
F      12752
M       6598
XNA        1
Name: monthly_income, dtype: int64

Distribution of missing values between education
BACHELOR'S DEGREE        23
Bachelor's Degree        25
PRIMARY EDUCATION         1
Primary Education         1
SECONDARY EDUCATION      67
SOME COLL

There is no unusal distribution of missing values of monthly_income (as well as symmetrically for days_employed) across all categorical columns in comparison to the distribution in the whole dataset. 

**Possible reasons for missing values in data**

Analisys prove that values were missed randomly, most likely due to technical reason while exporting or importing data in one step - the symmetrical loss of the values from 2174 rows hardly could happen in several iterations.   

**Intermediate conclusion**

The distribution in the whole dataset is the same that in the filtered table. It means that the values were missed randomly and most likely in one iteration. 


**Conclusions**

By filtering missing values in different columns we found out, that missing values are located symmetrically in two collumns: days_employed and monthly_income. There were found no other patterns for missing values. Missing values  

The data from days_employed are not our main variables of interest since they are not considered in the tasks of the poject.
Total days_employed column consist of impossible and missing values. Missing values in this column will be filled by calculation of maximum days_employed regarding age and gender of each customer. 

Missing values in monthly_income column will be replaced taking into considaration age and education of the customer.

Next step will be fixing letters' registers and removing possible duplications in education column, fixing impossible values in age column, removing artefacts in days_employed column, checking family_status and gender column.

## Data transformation

We will go through every column checking values for correct spelling and possible duplications.

In [11]:
# Let's see all values in education and education_id columns to check if and what spellings will need to be fixed
print(data_base['education'].unique())
print(data_base['education_id'].unique())

["bachelor's degree" 'secondary education' 'Secondary Education'
 'SECONDARY EDUCATION' "BACHELOR'S DEGREE" 'some college'
 'primary education' "Bachelor's Degree" 'SOME COLLEGE' 'Some College'
 'PRIMARY EDUCATION' 'Primary Education' 'Graduate Degree'
 'GRADUATE DEGREE' 'graduate degree']
[0 1 2 3 4]


Some values in the education column contain upper case letters or the combination of upper and a lower case letters.
All of them should be replaced with a lower case letters to get the understandings about the exact
amount of unique values in this column

In [12]:
# Fixing data - all values should be in lower case
data_base['education'] = data_base['education'].str.lower()

In [13]:
# Checking all the values in the column to make sure we fixed them
print(data_base['education'].unique())

# Cheking that all values correspondance to their id's
degrees = ["bachelor's degree",'secondary education', 'some college',
           'primary education', 'graduate degree']

# Define function which help us to check the one-to-one correspondsnce of 'education' to 'education_id'
def id_check(education):
    for i in education:
        print(i, data_base[data_base['education'] == i]['education_id'].unique())

#Calling function id_check
id_check(degrees)

["bachelor's degree" 'secondary education' 'some college'
 'primary education' 'graduate degree']
bachelor's degree [0]
secondary education [1]
some college [2]
primary education [3]
graduate degree [4]


Now everythinf looks fine and one-to-one correspndance established between education and
education_id values. We have 5 degrees in total in this category

Next step we will check the children column.

In [14]:
# Let's see the distribution of values in the `children` column
print(data_base['children'].value_counts())

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64


There are two issues in the children column: one is completely impossible negative number of children (equals to -1) and another one next to impossible (number of children equals to 20) regarding amount of cases, age of people, having 20 children and absence of entries between 5 and 20 children. Both issues are less then .5% in total data volume and distributed amonth the whole dataset. One can suggest that -1 value in fact belongs to "one child" and  20 value belongs to "two children" value and were caused by some technical errors.

In [15]:
# Fixing impossible value -1 to correct value 1
data_base['children'].replace(-1, 1, inplace=True)

# Fixing erroneus value 20 to correct value 2
data_base['children'].replace(20, 2, inplace=True)

In [16]:
# Checking the `children` column again to make sure it's all fixed
print(data_base['children'].value_counts())
print()
print(data_base.info())

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   age               21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   default           21525 non-null  int64  
 10  monthly_income    19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None


Now all values in children column looks normally and total amount of entries correponds to dataframe size.

Now we will work with values of `days_employed` column, which are either totaly impossible or partially incorrect. 

In [17]:
# Find problematic data in `days_employed`, if they exist, and calculate the percentage
print(data_base['days_employed'].describe())
negative = data_base[data_base['days_employed'] < 0]['gender'].count()
positive = data_base[data_base['days_employed'] >= 0]['gender'].count()
print()
print(f'Amount of negative values in column: ', negative)
print()
print(f'Amount of positive values in column: ', positive)
print((positive / (negative + positive))*100)

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

Amount of negative values in column:  15906

Amount of positive values in column:  3445
17.80269753501111


Let's see and evaluate the range of the positive values in the `days_employed` column.

In [18]:
# Creating dataframe, containing only positive values in days_employed column
is_null = data_base[data_base['days_employed'] > 0]
print(is_null.describe())

          children  days_employed          age  education_id  \
count  3445.000000    3445.000000  3445.000000   3445.000000   
mean      0.095791  365004.309916    59.124819      0.914659   
std       0.338400   21075.016396     7.580584      0.517103   
min       0.000000  328728.720605     0.000000      0.000000   
25%       0.000000  346639.413916    56.000000      1.000000   
50%       0.000000  365213.306266    60.000000      1.000000   
75%       0.000000  383246.444219    64.000000      1.000000   
max       4.000000  401755.400475    74.000000      4.000000   

       family_status_id      default  monthly_income  
count       3445.000000  3445.000000     3445.000000  
mean           0.984325     0.052830    21939.856893  
std            1.316071     0.223727    12838.753752  
min            0.000000     0.000000     3306.762000  
25%            0.000000     0.000000    13260.214000  
50%            0.000000     0.000000    18962.318000  
75%            2.000000     0.000000  

There are found 15906 entries with negative value, which is mostly likely a result of technical error and 3445 positive entries. All positive values should be also the result of some technical error, as their values are many times higher then the maximum amount for human beingn life (minimum positive value equals to 900 years). Thus all 100% of values in the column are impossible and may be caused by technical errors while importing or exporting them into final file we are working with. There are number of ways how to deal with this situation:
1. To drop this column totally since this data is not required in our project.
2. Try to fix impossible values.
There are also to ways of fixing data:
2.1. To replace all values with new values calculated in connection with gender, age and maximum possible work experience term or
2.2. To convert all negative values into positive ones and replace all impossible positive values according to clause 2.1.

Since work experince term is not required in our project, but it is usually the key information required by lenders for estimation of the credit score of the borrower it seems important to keep this column and try to replace impossible and missing values on the basis of other criteria and data from the dataset.

We will fix values in `days_employed` column in one step together with filling of missing values.

Now we will try to fix the impossible data in age column. There were entries with 0 age which is impossible for bank customer in our project.

In [19]:
print(data_base['age'].value_counts().sort_index())
print()
#Let's look for the distribution between genders in the whole and filtered dataframes

print(data_base.groupby('gender')['age'].count())
print()
print(data_base[data_base['age'] == 0]['gender'].value_counts())

0     101
19     14
20     51
21    111
22    183
23    254
24    264
25    357
26    408
27    493
28    503
29    545
30    540
31    560
32    510
33    581
34    603
35    617
36    555
37    537
38    598
39    573
40    609
41    607
42    597
43    513
44    547
45    497
46    475
47    480
48    538
49    508
50    514
51    448
52    484
53    459
54    479
55    443
56    487
57    460
58    461
59    444
60    377
61    355
62    352
63    269
64    265
65    194
66    183
67    167
68     99
69     85
70     65
71     58
72     33
73      8
74      6
75      1
Name: age, dtype: int64

gender
F      14236
M       7288
XNA        1
Name: age, dtype: int64

F    72
M    29
Name: gender, dtype: int64


There are found 101 entries with the 0 age value. Distribution among gender in the filtered dataframe equals to the distribution in the filtered one. This may be a human error, when sombody forgot to enter his age. The amount of 0 entries is very small and we can replace these values with mean value for 29 men (40 years) and 72 women (44 years) since the age is not a topic of our project, but 0 age person could not be subject of credit score. The mean value was chosen because we have pretty simmetric distribution withoгt extraordinary values in this column.

In [20]:
# Check the `dob_years` for suspicious values and count the percentage
mean_m = data_base[data_base['gender'] == 'M']['age'].mean()
mean_f = data_base[data_base['gender'] == 'F']['age'].mean()
print(f'Median men age: ', mean_m)
print(f'Median women age: ', mean_f)

Median men age:  40.993825466520306
Median women age:  44.47197246417533


All our age data are integers in this column so we will replace 0 values with `int(mean)` values for men and women respectively.

In [21]:
# Replacing 0 values in the age column
data_base.loc[(data_base['gender'] == 'M') & (data_base['age'] == 0), 'age'] = int(mean_m)                                                                                            
data_base.loc[(data_base['gender'] == 'F') &(data_base['age'] == 0), 'age'] = int(mean_f)                                                

In [22]:
# Checking the result
print(data_base['age'].describe())

count    21525.000000
mean        43.494448
std         12.218875
min         19.000000
25%         34.000000
50%         43.000000
75%         53.000000
max         75.000000
Name: age, dtype: float64


Now the age column containes regular data.

Now we are going to check the `family_status` column. We will check the unique values to see if there are any issues with texting (lower/upper letters for example) and create the function to check correspondance between `family_status` and `family_status_id`.

In [23]:
# Let's see the values for the column
status = data_base['family_status'].unique().tolist()
print(status)

# Defining function to check correspondance of status with stsus id
def id_check(education):
    for i in education:
        print(i, data_base[data_base['family_status'] == i]['family_status_id'].unique())

id_check(status)

['married', 'civil partnership', 'widow / widower', 'divorced', 'unmarried']
married [0]
civil partnership [1]
widow / widower [2]
divorced [3]
unmarried [4]


Hopefully in this column everything is fine and all values unambiguously corresponds to family_status_id.

Now let's check the gender column - unique values first.

In [24]:
# Let's see the values in the column
print(data_base['gender'].unique())

['F' 'M' 'XNA']


In addition to general male and female genders the XNA code was found.

In [25]:
# Let's see the entries with XNA value in 'gender'
print(data_base[data_base['gender']=='XNA'])

       children  days_employed  age     education  education_id  \
10701         0   -2358.600502   24  some college             2   

           family_status  family_status_id gender income_type  default  \
10701  civil partnership                 1    XNA    business        0   

       monthly_income          purpose  
10701       32624.825  buy real estate  


There is only one entry found, which could be either human error or the intentional entry since there are more 15 different gender's codes in use in Europe or USA, for example, and we do not know the region these data came from. In any case single entry will not affect on our analysis. 

Now we will look at the `income_type` column. We will first check the unique values and distribution among the whole dayaframe.

In [26]:
# Let's see the values in the column
print(data_base['income_type'].value_counts())

employee                       11119
business                        5085
retiree                         3856
civil servant                   1459
entrepreneur                       2
unemployed                         2
student                            1
paternity / maternity leave        1
Name: income_type, dtype: int64


We got another column with perfect data. No need to fix anything.

After fixing all data we now will try to find out are there any duplicates in our dataset.

In [27]:
# Checking duplicates
duplicated_rows = data_base.duplicated().sum()
print(f'Number of duplicated rows in the dataset: {duplicated_rows}')

Number of duplicated rows in the dataset: 71


In [28]:
# Address the duplicates, if they exist
data_base.drop_duplicates(inplace=True, ignore_index=True)

In [29]:
# Last check whether we have any duplicates

print(f'Number of duplicated rows in the dataset: {data_base.duplicated().sum()}')

Number of duplicated rows in the dataset: 0


Now our dataframe containes no duplications.

In [30]:
# Checking the size of the dataset that we now have after you first manipulations with it
print(data_base.info())

print(f'Percentage of lost data: {((duplicated_rows / raw_data)*100):.2%}')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   age               21454 non-null  int64  
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   default           21454 non-null  int64  
 10  monthly_income    19351 non-null  float64
 11  purpose           21454 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None
Percentage of lost data: 32.98%


Our new dataset consists of 21454 rows, which means that we had just 0.33% of duplicated entries in the original dataset. It now has 2103 entries with missing values which we have to replace with real data. 

# Working with missing values

### Restoring missing values in `monthly_income`

The income value basicly depends on the education, gender, work experience and type of income. In our dataset 17,8% of days_employed values are artefacts and will be replaced with calculated values. So we will take into consideration education, gender and income_type when filling missing values as identified factors. First we will need what identified factors will be more suitable for feeling missing values.

Values in monthle_income column have the wide range of valuues - from 3306 up to 362496. Whitin such conditions we will choose mean values within the certain groups for perlacing of missing values.

We will start by creating age_category column for the customers. By common practice there will be 10 year intervals in categories. 

In [31]:
# Let's write a function that calculates the age category

def assign_age_category(age):
    if age < 20:
        return '10-19'
    elif 20 <= age < 29:
        return '20-29'
    elif 30 <= age < 39:
        return '30-39'
    elif 40 <= age < 49:
        return '40-49'
    elif 50 <= age < 59:
        return '50-59'
    elif 60 <= age < 69:
        return '60-69'
    elif age >= 70:
        return '70+'  

In [32]:
# Test if the function works

def assign_age_category(age):
    if age < 20:
        return '10-19'
    elif 20 <= age < 30:
        return '20-29'
    elif 30 <= age < 40:
        return '30-39'
    elif 40 <= age < 50:
        return '40-49'
    elif 50 <= age < 60:
        return '50-59'
    elif 60 <= age < 70:
        return '60-69'
    elif age >= 70:
        return '70+'  

print(assign_age_category(34))
print(assign_age_category(49))
print(assign_age_category(58))

30-39
40-49
50-59


In [33]:
# Creating new column based on function
data_base['age_category'] = data_base['age'].apply(assign_age_category)

In [34]:
# Checking values in the new column
display(data_base.head(10))

children  days_employed  age            education  education_id  \
0         1   -8437.673028   42    bachelor's degree             0   
1         1   -4024.803754   36  secondary education             1   
2         0   -5623.422610   33  secondary education             1   
3         3   -4124.747207   32  secondary education             1   
4         0  340266.072047   53  secondary education             1   
5         0    -926.185831   27    bachelor's degree             0   
6         0   -2879.202052   43    bachelor's degree             0   
7         0    -152.779569   50  secondary education             1   
8         2   -6929.865299   35    bachelor's degree             0   
9         0   -2188.756445   41  secondary education             1   

       family_status  family_status_id gender income_type  default  \
0            married                 0      F    employee        0   
1            married                 0      F    employee        0   
2            married                 0      M    employee        0   
3            married                 0      M    employee        0   
4  civil partnership                 1      F     retiree        0   
5  civil partnership                 1      M    business        0   
6            married                 0      F    business        0   
7            married                 0      M    employee        0   
8  civil partnership                 1      F    employee        0   
9            married                 0      M    employee        0   

   monthly_income                              purpose age_category  
0       40620.102                purchase of the house        40-49  
1       17932.802                         car purchase        30-39  
2       23341.752                purchase of the house        30-39  
3       42820.568              supplementary education        30-39  
4       25378.572                    to have a wedding        50-59  
5       40922.170                purchase of the house        20-29  
6       38484.156                 housing transactions        40-49  
7       21731.829                            education        50-59  
8       15337.093                     having a wedding        30-39  
9       23108.150  purchase of the house for my family        40-49

We now will create the dataframe without missing values in monthly_income column for calculating values to replace missing values.

In [35]:
# Creating a table without missing values and printing a few of its rows to make sure it looks fine
data_base_wout_missing = data_base.dropna()
display(data_base_wout_missing.head())
print(data_base_wout_missing.info())

children  days_employed  age            education  education_id  \
0         1   -8437.673028   42    bachelor's degree             0   
1         1   -4024.803754   36  secondary education             1   
2         0   -5623.422610   33  secondary education             1   
3         3   -4124.747207   32  secondary education             1   
4         0  340266.072047   53  secondary education             1   

       family_status  family_status_id gender income_type  default  \
0            married                 0      F    employee        0   
1            married                 0      F    employee        0   
2            married                 0      M    employee        0   
3            married                 0      M    employee        0   
4  civil partnership                 1      F     retiree        0   

   monthly_income                  purpose age_category  
0       40620.102    purchase of the house        40-49  
1       17932.802             car purchase        30-39  
2       23341.752    purchase of the house        30-39  
3       42820.568  supplementary education        30-39  
4       25378.572        to have a wedding        50-59

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19351 entries, 0 to 21453
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          19351 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   age               19351 non-null  int64  
 3   education         19351 non-null  object 
 4   education_id      19351 non-null  int64  
 5   family_status     19351 non-null  object 
 6   family_status_id  19351 non-null  int64  
 7   gender            19351 non-null  object 
 8   income_type       19351 non-null  object 
 9   default           19351 non-null  int64  
 10  monthly_income    19351 non-null  float64
 11  purpose           19351 non-null  object 
 12  age_category      19351 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB
None


The dataframe consists of 19351 rows with no missing values.

Now we can check wich identified factors abest fit our goal.

In [36]:
# Median table for age and gender

pivot_table_monthly_income = data_base_wout_missing.pivot_table(
                                                                index='age_category',
                                                                columns = 'gender',
                                                                values = 'monthly_income',
                                                                aggfunc='median')
display(pivot_table_monthly_income)

gender                 F           M        XNA
age_category                                   
10-19         17954.8410  14700.2810        NaN
20-29         20451.7825  26008.6940  32624.825
30-39         22328.9650  28161.4955        NaN
40-49         22672.9320  28401.6975        NaN
50-59         21028.0100  25703.8170        NaN
60-69         19431.7180  21472.5740        NaN
70+           17864.7065  20526.6140        NaN

In [37]:
# Median table for income_type and gender 

pivot_table_total_income_type_median = data_base_wout_missing.pivot_table(
                                                                        index='income_type',
                                                                        columns = 'gender',
                                                                        values = 'monthly_income',
                                                                        aggfunc='median')
display(pivot_table_total_income_type_median)

gender                                F           M        XNA
income_type                                                   
business                     25731.3245  31491.0080  32624.825
civil servant                21917.1980  29754.3915        NaN
employee                     20898.4980  25945.7880        NaN
entrepreneur                 79866.1030         NaN        NaN
paternity / maternity leave   8612.6610         NaN        NaN
retiree                      18529.2465  20918.3620        NaN
student                             NaN  15712.2600        NaN
unemployed                   32435.6020   9593.1190        NaN

In [38]:
# Median table for education and gender

pivot_table_monthly_income_edu_median = data_base_wout_missing.pivot_table(
                                                                            index= 'education',
                                                                            columns = 'gender',
                                                                            values = 'monthly_income',
                                                                            aggfunc='median')

display(pivot_table_monthly_income_edu_median)

gender                        F           M        XNA
education                                             
bachelor's degree    26063.4715  32675.8355        NaN
graduate degree      29345.3940  25161.5835        NaN
primary education    17223.9615  21204.0860        NaN
secondary education  20101.2700  25435.5815        NaN
some college         22836.0820  29973.6640  32624.825

Among all tables above the most accurate data, in our opinion, contained in the Median table for education and gender, which contains no NaN values (XNA code equals to just one entry, which contains monthly_income value).

In [39]:
#  Defining a function that we will use for filling in missing values
def assign_missing_m_i(data_base):
    education = data_base['education']
    gender = data_base['gender']
    if pd.isnull(data_base['monthly_income']):
        return pivot_table_monthly_income_edu_median.loc[education, gender]
    else:
        return data_base['monthly_income']

In [40]:
# Checking if it works
print(assign_missing_m_i(data_base.loc[12]))
print(assign_missing_m_i(data_base.loc[0]))
print(assign_missing_m_i(data_base.loc[29]))
print(assign_missing_m_i(data_base.loc[3]))
print(assign_missing_m_i(data_base.loc[21424]))
print(assign_missing_m_i(data_base.loc[21431]))
print(assign_missing_m_i(data_base.loc[21439]))
print(assign_missing_m_i(data_base.loc[21426]))


25435.5815
40620.102
20101.27
42820.568
20101.27
20101.27
20101.27
26063.4715


In [41]:
# Appling it to every row
data_base['monthly_income'] = data_base.apply(assign_missing_m_i, axis=1)

In [42]:
# Checking if we got any errors
print(f'Total amount of missing values in monthly_income column is {data_base["monthly_income"].isna().sum()}')

Total amount of missing values in monthly_income column is 0


Now we have no missing values in the monthly income column and we shall check the number of entries in monthly income column.

In [43]:
# Checking the number of entries in the columns
print(data_base.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   age               21454 non-null  int64  
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   default           21454 non-null  int64  
 10  monthly_income    21454 non-null  float64
 11  purpose           21454 non-null  object 
 12  age_category      21454 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB
None


Total amount of entries in monthly_income column corresponds to our dataframe size.

###  Restoring values in `days_employed`

Since we have found, that all values in days_employed column are impossible, we will restore values in this column based on the following principles. We assume maximum retire age for men as 60 years and 55 years fo women. We assume that starting age for the purpose of our research should be lower then minimum age from age column of our dataset and equals to 16 years.

All negative values in the column shoud be converted to positive ones with the same absolute value, while all positive values will be treated as follows: if age is less then retire age, days_employed value will be calculated based on the gender and actual age, assuming that client starteв to work at the start age. For ages higher then retire age the employed_age value will be equal to the maximim work experience term, taking gender into considaration.  


In [44]:
# Let's write a function that assigns values in the days_employed column

men_retire_age = 60
start_age = 16
women_retire_age = 55

max_work_experience_m = (men_retire_age - start_age) * 365
max_work_experience_f = (women_retire_age - start_age) * 365   

print(max_work_experience_m)
print(max_work_experience_f)

def work_experience_fill(data_base):
    if data_base['days_employed'] < 0:
        return abs(data_base['days_employed'])       
    elif pd.isnull(data_base['days_employed']):
        if data_base['gender'] == 'M':
            if data_base['age'] < men_retire_age:
                return (data_base['age'] - start_age) * 365
            return max_work_experience_m
        elif data_base['gender'] == 'F':
            if data_base['age'] < women_retire_age:
                return (data_base['age'] - start_age) * 365
            return max_work_experience_f        
    else:
        if data_base['gender'] == 'M' and data_base['days_employed'] > max_work_experience_m:
            if data_base['age'] < men_retire_age:
                return (data_base['age'] - start_age) * 365
            return max_work_experience_m
        elif data_base['gender'] == 'F' and data_base['days_employed'] > max_work_experience_f:
            if data_base['age'] < women_retire_age:
                return (data_base['age'] - start_age) * 365
            return max_work_experience_f

16060
14235


Let's check if the function works correct for the values from our dataset.

In [45]:
# Checking that the function works

print(work_experience_fill(data_base.loc[1]))
print(work_experience_fill(data_base.loc[26]))
print(work_experience_fill(data_base.loc[4]))
print(work_experience_fill(data_base.loc[25]))


4024.803753850451
9125
13505
16060


The function works correct and it may be applied to all rows of the dataframe.

In [46]:
# Apply function to the every row of 'days_employed' column
data_base['days_employed'] = data_base.apply(work_experience_fill, axis=1)

In [47]:
# Check if function worked
print(data_base['days_employed'].isnull().sum())
print(data_base.isnull().sum())

0
children            0
days_employed       0
age                 0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
default             0
monthly_income      0
purpose             0
age_category        0
dtype: int64


Now all missing and impossible values in the dataset are fixed. We can change datatype of days_employed and monthly_income columns to int.

In [48]:
# Changing datatype
data_base = data_base.astype({'days_employed' : 'int', 'monthly_income' : 'int'})

We now need to check that the size of the dataframe and general information.

In [49]:
# Check the entries in all columns - make sure we fixed all missing values
print(data_base.info())
print()
print(data_base.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21454 non-null  int64 
 1   days_employed     21454 non-null  int64 
 2   age               21454 non-null  int64 
 3   education         21454 non-null  object
 4   education_id      21454 non-null  int64 
 5   family_status     21454 non-null  object
 6   family_status_id  21454 non-null  int64 
 7   gender            21454 non-null  object
 8   income_type       21454 non-null  object
 9   default           21454 non-null  int64 
 10  monthly_income    21454 non-null  int64 
 11  purpose           21454 non-null  object
 12  age_category      21454 non-null  object
dtypes: int64(7), object(6)
memory usage: 2.1+ MB
None

           children  days_employed           age  education_id  \
count  21454.000000   21454.000000  21454.000000  21454.000000   
mean       0.

All columns of the dataframe now have correct values and no missing values.

## Categorization of data

The questions that we have to answer in this project requares us to categorize our data since nearly all columns have two many unique values for correct analysis. We need to assing the appropriate categories to: `children` column, `family_status` column, `income_level` column and `purpose` column
    
Categorizations process we will start from text data following by numeric data columns. 

First of all we will check the ubique values in all columns (except monthly_income, which has too many unique values)

In [50]:
# Check the unique values
print(data_base['family_status'].value_counts())
print()
print(data_base['purpose'].value_counts())
print()
print(data_base['children'].value_counts())

married              12339
civil partnership     4151
unmarried             2810
divorced              1195
widow / widower        959
Name: family_status, dtype: int64

wedding ceremony                            791
having a wedding                            768
to have a wedding                           765
real estate transactions                    675
buy commercial real estate                  661
housing transactions                        652
buying property for renting out             651
transactions with commercial real estate    650
purchase of the house                       646
housing                                     646
purchase of the house for my family         638
construction of own property                635
property                                    633
transactions with my real estate            627
building a real estate                      624
buy real estate                             621
purchase of my own house                    620
building a pro


# Martial status

Martial status (family_status column) has only 5 unique values which hardly could be categorized and could serve as categories as they are.

# Purpose

Since the column purpose of the loan seems to have the free type of text values (in comparison to family_status and education columns) it may contain one and the same purpose expressed in different words, for example all purpose values wich contain the word "wedding" are actually refers to the loan for wedding preparations. This is also true for all purpose values, which contains "car" - all of them are actually car loans. Education category includes all values, which contain words education (in all forms) and univercity. 

The situation with real estate loan purpose values is a little bit more complicated. We can define 4 major categories of loan purposes: 
financing of residential real estate (purchase of the house for my family, purchase of my own house, buy residential real estate); 
financing of commercial real estate (buy commercial real estate, buying property for renting out, transactions with commercial real estate); 
financing of construction and reconstruction/rennovation of real estate (construction of own property, building a property, building a real estate, housing renovation) and 
other purposes, which are not clearly defined (housing, real estate transactions, transactions with my real estate, etc.)  

In [51]:
# Let's pregroup purpose column for the values referring to real estate
data_base['purpose'].replace(['buy commercial real estate', 'buying property for renting out',
                                                     'transactions with commercial real estate'], 
                                                     'commercial real estate', inplace=True) 

data_base['purpose'].replace(['construction of own property', 'building a property', 
                                                     'building a real estate', 'housing renovation'], 
                                                     'construction and rennovation', inplace=True)

data_base['purpose'].replace(['purchase of the house for my family', 'purchase of my own house', 
                                                     'buy residential real estate'], 'residential real estate', 
                                                    inplace=True)       

data_base['purpose'].replace(['real estate transactions', 'housing transactions', 
                                                     'purchase of the house', 'housing', 'property', 
                                                     'transactions with my real estate', 'buy real estate'], 
                                                     'other real estate', inplace=True)

In [52]:
# Let's check the pregroupped purposes
print(data_base['purpose'].value_counts())

other real estate                   4500
construction and rennovation        2485
commercial real estate              1962
residential real estate             1864
wedding ceremony                     791
having a wedding                     768
to have a wedding                    765
buying my own car                    505
going to university                  496
car                                  494
second-hand car purchase             486
cars                                 478
to own a car                         478
buying a second-hand car             478
to buy a car                         471
car purchase                         461
supplementary education              460
purchase of a car                    455
university education                 452
education                            447
to get a supplementary education     446
getting an education                 442
profile education                    436
getting higher education             426
to become educat

Now the real estate section of the purpose column looks more accurate and more informative and we can categorize the rest values into following categorises: car loans, wedding and education.

In [53]:
# Let's write a function for assigning purpose to major caterories
def assign_purpose_category(purpose):
    if 'car' in purpose:
        return 'car loan'
    elif 'wedding' in purpose:
        return 'wedding'
    elif 'educat' in purpose or 'university' in purpose:
        return 'education'
    elif 'construction and rennovation' in purpose:
        return 'construction and rennovation'    
    elif 'commercial' in purpose:
        return 'commercial real estate'    
    elif 'residential' in purpose:
        return 'residentail real estate'        
    elif 'other real estate'in purpose:
        return 'other real estate'
    else:
        return 'other'

In [54]:
# Check if it works

print(assign_purpose_category('buy residential real estate'))
print(assign_purpose_category('for a wedding'))
print(assign_purpose_category('buy a second hand car'))
print(assign_purpose_category('buy a commercial real estate'))
print(assign_purpose_category('for university'))
print(assign_purpose_category('construction and rennovation'))
print(assign_purpose_category('other real estate'))
print(assign_purpose_category('to get a supplementary education'))
print(assign_purpose_category('cars'))

residentail real estate
wedding
car loan
commercial real estate
education
construction and rennovation
other real estate
education
car loan


In [55]:
# Appling to each row
data_base['purpose_category'] = data_base['purpose'].apply(assign_purpose_category)

In [56]:
# Chek total info
print(data_base.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21454 non-null  int64 
 1   days_employed     21454 non-null  int64 
 2   age               21454 non-null  int64 
 3   education         21454 non-null  object
 4   education_id      21454 non-null  int64 
 5   family_status     21454 non-null  object
 6   family_status_id  21454 non-null  int64 
 7   gender            21454 non-null  object
 8   income_type       21454 non-null  object
 9   default           21454 non-null  int64 
 10  monthly_income    21454 non-null  int64 
 11  purpose           21454 non-null  object
 12  age_category      21454 non-null  object
 13  purpose_category  21454 non-null  object
dtypes: int64(7), object(7)
memory usage: 2.3+ MB
None


All columns in new dataframe are of the same size.

In [57]:
print(data_base['purpose_category'].value_counts())

other real estate               4500
car loan                        4306
education                       4013
construction and rennovation    2485
wedding                         2324
commercial real estate          1962
residentail real estate         1864
Name: purpose_category, dtype: int64


Now we have the appropriate amount of categories which covers all major financial activities of the customers: morgage, construction and rennovation of real estate, commercial real estate financing, wedddings preparations, financing of education and car loans.  

# Income level 




Typical categorization of income levels refers to 3 levels - low, middle and high. Since we do not know exactly the region and the currency represented in our selection we can rely on the definition of the middle class income level boundaries, thus defining simaltaneously other two levels: low level is less, then lower boundary and high level is higher then upper bondary of middle level. 

According to the most common economic theories the middle class income qualifies between 75% and 125% of the median income in the region. So we can calculate the median of the `monthly_income` column values and define the categories' values.

In [58]:
# Calculating monthly_income column median
print(data_base['monthly_income'].median())

23180.0


In [59]:
# Let's write a function to categorize the income level 
middle_class_low_boundary = 0.75
middle_class_high_boundary = 1.25

def assign_monthly_income_category(monthly_income):
    
    middle_class_low = data_base['monthly_income'].median() * middle_class_low_boundary
    middle_class_high = data_base['monthly_income'].median() * middle_class_high_boundary

    if monthly_income <= middle_class_low:
        return 'low'
    elif middle_class_low < monthly_income <= middle_class_high:
        return "middle"
    else:
        return "high"

Now we need to check how the function works.

In [60]:
# Checking if it works
print(assign_monthly_income_category(5000))
print(assign_monthly_income_category(25000))
print(assign_monthly_income_category(75000))

low
middle
high


The function seems to work correctly and we can apply it to the whole monthly_income column and create a new column monthly_income_category.

In [61]:
# Create a column with the categories and count the values for them
data_base['monthly_income_category'] = data_base['monthly_income'].apply(assign_monthly_income_category)

In [62]:
# Cheking the distribution 
print(data_base['monthly_income_category'].value_counts())

middle    9406
high      6545
low       5503
Name: monthly_income_category, dtype: int64


In [63]:
# Checking global info 
print(data_base.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   children                 21454 non-null  int64 
 1   days_employed            21454 non-null  int64 
 2   age                      21454 non-null  int64 
 3   education                21454 non-null  object
 4   education_id             21454 non-null  int64 
 5   family_status            21454 non-null  object
 6   family_status_id         21454 non-null  int64 
 7   gender                   21454 non-null  object
 8   income_type              21454 non-null  object
 9   default                  21454 non-null  int64 
 10  monthly_income           21454 non-null  int64 
 11  purpose                  21454 non-null  object
 12  age_category             21454 non-null  object
 13  purpose_category         21454 non-null  object
 14  monthly_income_category  21454 non-nul

The distribution of the `monthly_income` between categories in our dataset and the size of the new column and the dataframe looks fine.

# Children

Regarding the last column with values for categorization, we can devide it into 2 categories: one whith children and another one - without children. But since we have a range of data regarding children (for those who have children) from 1 to 5, it seems interesting to analize how the amount of children affect the default ratio. That's why we will create 3 categories:
    no children
    one child 
    many children 

For this purpose we will create a special funcrion 

In [64]:
# Creating function for categorizing into different numerical groups based on ranges
def assign_children_category(children):
    if children == 0:
        return 'no children'
    elif children == 1:
        return 'one child'
    else: return 'many children'

In [65]:
# Checking if it works
print(assign_children_category(0))
print(assign_children_category(1))
print(assign_children_category(5))

no children
one child
many children


The functions works accurate and we can apply it to the whole column. Then we check the distribution and the size of the new dataframe.

In [66]:
# Creating column with categories
data_base['children_category'] = data_base['children'].apply(assign_children_category)

In [67]:
# Counting each categories values to see the distribution
print(data_base['children_category'].value_counts())

no children      14091
one child         4855
many children     2508
Name: children_category, dtype: int64


In [68]:
# Cheking general info
print(data_base.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   children                 21454 non-null  int64 
 1   days_employed            21454 non-null  int64 
 2   age                      21454 non-null  int64 
 3   education                21454 non-null  object
 4   education_id             21454 non-null  int64 
 5   family_status            21454 non-null  object
 6   family_status_id         21454 non-null  int64 
 7   gender                   21454 non-null  object
 8   income_type              21454 non-null  object
 9   default                  21454 non-null  int64 
 10  monthly_income           21454 non-null  int64 
 11  purpose                  21454 non-null  object
 12  age_category             21454 non-null  object
 13  purpose_category         21454 non-null  object
 14  monthly_income_category  21454 non-nul

The distribution and dataframe size look fine.

We have categorized data in 3 columns - one columns, containig text data (categorical/qualitive data) - purpose column - and two columns containing numeric data (quantative data) - number of children (children) and total income (monthly_income). Family_status column contains 5 values with one to one correspondance to family_status_id values, which can be used as a categories as they are. Now we can check if hypotheses we need to prove correct or not.   

## Checking the Hypotheses


**Is there a correlation between having children and paying back on time?**

In [69]:
# Checking the children data and paying back on time
children_category_default = data_base.groupby(['children_category','default']).agg({'default':["count"]})
display(children_category_default)
print()

# Calculating default-rate based on the number of children
def children_category_default_ratio(df, max_index, category_names):
    count = 0
    for i in range(1, max_index + 1, 2):
        total = df.iloc[i] + df.iloc[i-1]
        ratio  = (df.iloc[i] / total)[0]
        print(f'Default ratio for customer from {category_names[count]} category is {ratio:.2%}')
        count += 1

# Applying function        
children_category_default_ratio(children_category_default, 5, ['many children', 'no children', 'one child'])

default
                            count
children_category default        
many children     0          2275
                  1           233
no children       0         13028
                  1          1063
one child         0          4410
                  1           445


Default ratio for customer from many children category is 9.29%
Default ratio for customer from no children category is 7.54%
Default ratio for customer from one child category is 9.17%


**Conclusion**

One can see the affect of childrens' factor to the default ratio of the customer: customers without children perform the lowest deafault ratio of 7.54%, while for the customers having one child it's much higher - 9.17%. This ratio is even higher for customers having more then one child - 9.29%.


**Is there a correlation between family status and paying back on time?**

In [70]:
# Check the family status data and paying back on time
family_status_default = data_base.groupby(['family_status','default']).agg({'default':["count"]})
display(family_status_default)
print()

# Calculating default-rate based on family status

def family_status_default_ratio(df, max_index, category_names):
    count = 0
    for i in range(1, max_index + 1, 2):
        total = df.iloc[i] + df.iloc[i-1]
        ratio  = (df.iloc[i] / total)[0]
        print(f'Default ratio for customer from {category_names[count]} is {ratio:.2%}')
        count += 1
# Applying function        
family_status_default_ratio(family_status_default, 9, ['civil partnership', 'divorced', 'married', 
                                                       'unmarried', 'widow / widower'])


default
                            count
family_status     default        
civil partnership 0          3763
                  1           388
divorced          0          1110
                  1            85
married           0         11408
                  1           931
unmarried         0          2536
                  1           274
widow / widower   0           896
                  1            63


Default ratio for customer from civil partnership is 9.35%
Default ratio for customer from divorced is 7.11%
Default ratio for customer from married is 7.55%
Default ratio for customer from unmarried is 9.75%
Default ratio for customer from widow / widower is 6.57%


**Conclusion**

There is a strong correlation betweуn martial status and paying back on time. The highest score in this category belongs to widows/widowers - 6.57% default ratio, while unmarried customers shows the worst result of 9.75% default ratio. Between them divodced customers with 7.11%, married customers with 7.55% and civil partnerships with 9.35% are located. 

**Is there a correlation between income level and paying back on time?**

In [71]:
# Check the income level data and paying back on time
monthly_income_category_default = data_base.groupby(['monthly_income_category','default']).agg({'default':["count"]})
display(monthly_income_category_default)
print()

# Calculating default-rate based on income level

def monthly_income_category_default_ratio(df, max_index, category_names):
    count = 0
    for i in range(1, max_index + 1, 2):
        total = df.iloc[i] + df.iloc[i-1]
        ratio  = (df.iloc[i] / total)[0]
        print(f'Default ratio for customer from {category_names[count]} is {ratio:.2%}')
        count += 1

# Applying function        
monthly_income_category_default_ratio(monthly_income_category_default, 5, ['high', 'low', 'middle'])

default
                                  count
monthly_income_category default        
high                    0          6062
                        1           483
low                     0          5064
                        1           439
middle                  0          8587
                        1           819


Default ratio for customer from high is 7.38%
Default ratio for customer from low is 7.98%
Default ratio for customer from middle is 8.71%


**Conclusion**

Our reseach shows that customers with middle level income performs the higher level of default ratio of 8.71%, while the minimum level of default of 7.38% refers to the customers with hihg level of income. Customers with low income shows middle default rate of 7.98%

**How does credit purpose affect the default rate?**

In [72]:
# Check the percentages for default rate for each credit purpose and analyze them

purpose_category_default = data_base.groupby(['purpose_category','default']).agg({'default':["count"]})
display(purpose_category_default)
print()

# Defining function
def purpose_category_default_ratio(df, max_index, category_names):
    count = 0
    for i in range(1, max_index + 1, 2):
        total = df.iloc[i] + df.iloc[i-1]
        ratio  = (df.iloc[i] / total)[0]
        print(f'Default ratio for customer from {category_names[count]} is {ratio:.2%}')
        count += 1

# Applying function        
purpose_category_default_ratio(purpose_category_default, 13, ['car loan', 'commercial real estate', 
                                                              'construction and rennovation', 'education', 
                                                              'other real estate', 
                                                              'residential real estate', 'wedding'])

default
                                       count
purpose_category             default        
car loan                     0          3903
                             1           403
commercial real estate       0          1811
                             1           151
construction and rennovation 0          2306
                             1           179
education                    0          3643
                             1           370
other real estate            0          4168
                             1           332
residentail real estate      0          1744
                             1           120
wedding                      0          2138
                             1           186


Default ratio for customer from car loan is 9.36%
Default ratio for customer from commercial real estate is 7.70%
Default ratio for customer from construction and rennovation is 7.20%
Default ratio for customer from education is 9.22%
Default ratio for customer from other real estate is 7.38%
Default ratio for customer from residential real estate is 6.44%
Default ratio for customer from wedding is 8.00%


**Conclusion**

Based on the dataset provided we can conclude, that different purpose of credit shows very different default ratios varies from 6.44% for morgage and up to 9.36% for car loans. In general all real estate related loans, including those for construction and rennovation (7.20%), commercial real estate (7.70%) and othe real estate related loans (7.38%) shows lower default levels in comparison with loans for education (9.22%) and wedding preparations (8.00%).  


# General Conclusion 

During the analysis of the dataset provided we had to check data integrity, find impossible and missing values and replace them with the most accurate values. There was established and proved symmetric character of missing values in two columns - monthly_income and days-employed in total amount of 10% of the whole dataframe. The data in days_employed column were found totally impossible with 17.8% of positive values exceeding all human beings abilities and the rest consisted of negative values. There were esablished some impossible values in age column and upper case texting in education column.  The purpose column contained a lot of entries, where the same purpose was texted in different word forms and/or expressions (there were potentially a number of duplicated rows regarding this issue, but to establish theese duplicates correctly we need more external information). Three columns in this dataset contained perfect data (education, education_id and income_type_id columns).

In order to perform our analysis we need to decide the right methods to fix all the above mentioned issues. The easiest issue with upper case letters in education column was resolved with pandas method. In the age column 0 age values were replaced with mean ages for the certain gender. In the children column negative values were replaced by positive ones with the same absolute value and the extraordinary erroneus value of 20 children was reduced to 2. 

71 duplicates were found after fixing issues in education, age and columns.

Regarding two most serious issues with missing values the following conclusions were made. For days_employed values all negative entries were replaced by positive ones with the same absolute value. All impossible values were replaced with a special function for the calculated values with regard to the gender and age of the customer. Since data from days_employed column do not present in the project goals thees replacements do not affect our conclusions.

For monthly_income column the descision was to replace missing values to the medians of the respective categories filtered on education and gender.



1. One can see the affect of childrens' factor to the default ratio of the customer: customers without children perform the lowest deafault ratio of 7.54%, while for the customers having one child it's much higher - 9.17%. This ratio is even higher for customers having more then one child - 9.29%.

2. There is a strong correlation betweуn martial status and paying back on time. The highest score in this category belongs to widows/widowers - 6.57% default ratio, while unmarried customers shows the worst result of 9.75% default ratio. Between them divodced customers with 7.11%, married customers with 7.55% and civil partnerships with 9.35% are located. 

3. Our reseach shows that customers with middle level income performs the higher level of default ratio of 8.71%, while the minimum level of default of 7.38% refers to the customers with hihg level of income. Customers with low income shows middle default rate of 7.98%

4. Based on the dataset provided we can conclude, that different purpose of credit shows very different default ratios varies from 6.44% for morgage and up to 9.36% for car loans. In general all real estate related loans, including those for construction and rennovation (7.20%), commercial real estate (7.70%) and othe real estate related loans (7.38%) shows lower default levels in comparison with loans for education (9.22%) and wedding preparations (8.00%).  
